### ISSUE TITLE
How to use both CPU and GPU for training in distributed training

### ISSUE THREAD

**Issue by limapedro** [tensorflow/tensorflow#40767]
Hi, I'd like to know how to use both CPU and GPU for training, this is how my solution looks like, but it only uses the GPU.

strategy = tf.distribute.MirroredStrategy(devices=['CPU', 'DML'])

If I do,

with tf.device('CPU'):
    model.fit(....)

This uses the CPU, how can I use both to speed up things a little.

**Comment by nikitamaia** [comment]
Hi @limapedro, if you only have 1 GPU you're probably best off not using distributed training and just running your code with your GPU. You won't be able to run MirroredStrategy with 1 GPU and 1 CPU. Think of it this way, in MirroredStrategy your model is replicated on each device. Each device receives a portion of the input data batch and computes the gradient. The gradients are then aggregated across all of your devices, and this new reduced gradient is what is used by each device to update the parameters of your model. Then this process happens again, taking another step in your training loop. If you were to use MirroredStrategy with 1 GPU and 1 CPU in sync, then your GPU would just be waiting around each step after the backwards pass for the CPU to finish so that the gradient could be updated. And your CPU would bottleneck.

Hope that explanation helps. But it doesn't really sound like you have a need for distributed training in this case if you only have one GPU. TensorFlow code, and tf.keras models will transparently run on a single GPU with no code changes required.

**Comment by limapedro** [comment]
@nikitamaia Thanks for replying, I was looking for to this because my GPU isn't much faster than my CPU on tensorflow, so using both would improve training times a bit, my device is DML one, from tensorflow-directml.

**Comment by nikitamaia** [comment]
I'm not familiar with DML, so I can't speak to that. Though it is surprising that training with GPU isn't much faster than training on CPU. However, as I mentioned earlier trying to use MirroredStrategy with 1 CPU and 1 GPU will not improve your training time.

**Comment by limapedro** [comment]
@nikitamaia Yeahm I using Vega 11, AMD integrated graphics, It is about 2x times faster running on Vega 11 versus my Ryzen CPU, I expect DML to become more popular in the following months. both for training and inference.
